bluetooth: host: convert conn address on identity resolution

kapi-no opened the pull request:
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/61028

Fix an issue in the CCC configuration when the Peripheral device maintains two or more distinct connections on different identities with the same peer. The issue occurs when the local device performs the pairing and the bonding procedure on a connection associated with one of the Bluetooth identities. During the identity resolution, the peer address field in the CCC descriptor is converted from the RPA-type address to the Identity Address. However, the destination address on the remaining connection objects associated with other Bluetooth identities is not converted. Due to this, their CCC configuration is reset and GATT indications and notification fail to be sent even if the Central device subscribed to them.

Added necessary code to iterate over all connection objects during the identity resolution phase and aligned their destination address from the RPA-type to the Identity Address.

kapi-no committed:
bluetooth: host: convert conn address on identity resolution

Fix an issue in the CCC configuration when the Peripheral device
maintains two or more distinct connections on different identities
with the same peer. The issue occurs when the local device performs
the pairing and the bonding procedure on a connection associated
with one of the Bluetooth identities. During the identity resolution,
the peer address field in the CCC descriptor is converted from the
RPA-type address to the Identity Address. However, the destination
address on the remaining connection objects associated with other
Bluetooth identities is not converted. Due to this, their CCC
configuration is reset and GATT indications and notification fail
to be sent even if the Central device subscribed to them.

Added necessary code to iterate over all connection objects during
the identity resolution phase and aligned their destination address
from the RPA-type to the Identity Address.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3874,8 +3888,15 @@ static uint8_t smp_ident_addr_info(struct bt_smp *smp, struct net_buf *buf)
 			 * present before ie. due to re-pairing.
 			 */
 			if (!bt_addr_le_is_identity(&conn->le.dst)) {
+				struct addr_match addr_match = {
```
Could you add a bsim test that exercises this part of the code? I'm fine with the current changes as long as they are tested.

kapi-no replied:
Added a BabbleSim test as you requested

kapi-no committed:
bluetooth: tests: add bsim test for identity address update

Added a BabbleSim test to check the update of the destination address
of all active connections during the identity resolution operation.
All connection objects associated with the same private peer address
should convert it to the identity address.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
> > when the Peripheral device maintains two or more distinct connections on different identities with the same peer
> 
> @kapi-no Could you elaborate on the use-case? I'm not sure this should be supported at all, given the headaches this kind of usage might bring.

The use case is one of the asset-tracking protocols based on Bluetooth - called Apple Find My, which implements its own Privacy solution instead of using a standardized Bluetooth Privacy feature. The Find My Identity needs to manage its advertising address (using the `bt_id_reset` function in the context of Zephyr API). Unfortunately, I cannot share any details regarding the protocol design as it is closed-source.

When such an asset-tracking protocol is combined with the main Bluetooth functionality, each functionality needs to have its own separate identity. Let's use an example of the LE Audio headphones with the location tracking feature (Apple Find My). In this case, separate identities prevent the location-tracking functionality from resetting the bonding information of the LE audio functionality during the advertising address updates with `bt_id_reset` function.

This use case is kind of similar to the HW approach where typically product has two different chipsets that require dedicated connections. In this case, however, we have one chip but the separation happens at the software level thanks to the dedicated Bluetooth identities.

sjanc commented:
Hi,

How does this comply with Core Spec 5.4, Vol 6, Part B, 1.1.1 "Permitted state and role combinations" ?

"The Link Layer may optionally support multiple state machines. If it does
support multiple state machines, then any combination of states and roles may
be supported. In particular, the Link Layer may be in the Connection State
multiple times with any mix of Central Role and Peripheral Role. **However, any
two devices shall not have more than one connection between them**."

kapi-no commented:
> Hi,
> 
> How does this comply with Core Spec 5.4, Vol 6, Part B, 1.1.1 "Permitted state and role combinations" ?
> 
> "The Link Layer may optionally support multiple state machines. If it does support multiple state machines, then any combination of states and roles may be supported. In particular, the Link Layer may be in the Connection State multiple times with any mix of Central Role and Peripheral Role. **However, any two devices shall not have more than one connection between them**."

Well, the term device is not defined in the specification so it leaves some room for interpretation. If we understand the term device as device identity (identified by the unique Identity Address exchanged during the pairing), then this change is acceptable.

However, this PR does not attempt to change anything at the Link Layer level. It simply supports the Link Layer implementations (e.g. Zephyr LL, Softdevice Controller LL from the nRF Connect SDK) that interpret the word device as outlined by me in the previous section. If any of the LL implementations choose to discard the second connection with the same peer, my code won't do any harm to comply with this requirement.

sjanc commented:
Hi,

While I understand multiple ID usecase I have trouble understanding this patch.

Why all matching connections being updated on resolution and not just one that pairing was performed for?
What is more for those other conn bt_conn_identity_resolved() isn't called. 

I looked at test and I'm even more confused as this looks like pairing is expected to happen only on one connection while dst is expected to resolv on all RPA maching?  Are all updated connection now suppose to use same keys?

What if rogue device copy old peer-RPA and connect, will it be able to update dst (ID?) addr of other (yet to be paired) device?  This will prevent that device from properly notify on own addr resolution on pairing?

alwa-nordic commented:
> What if rogue device copy old peer-RPA and connect, will it be able to update dst (ID?) addr of other (yet to be paired) device?

This is a very good question. I'll dismiss my review until this is answered.

jori-nordic commented:
After a second re-read of the bug report, I also tend to believe this is the wrong fix.

To me, it rather seems that it's the identities feature that is currently broken in GATT. Ideally, the CCC array would be virtualized for each identity. Since we don't have tests with that weird use-case (same peer, diff id) we never caught it.

@alwa-nordic and I have been thinking for a while that the identities abstraction is not well-suited to HCI, especially regarding privacy. And here we also see that it was kind of half-assed, as it should ideally virtualize everything about the host, eg. the name, the whole ATT table (at least the GAP characteristic which sets the bondable mode), etc.. for every identity.

It would be great if we had some usage data about identities, ie what people are actually using it for so we can figure out a better system.

kapi-no commented on tests/bsim/bluetooth/host/security/id_addr_update/src/peripheral.c:
```diff
@@ -0,0 +1,79 @@
+	struct bt_conn *conn_a;
+	struct bt_conn *conn_b;
+
+	/* Create two identtities that will simulatenously connect with the same cenral peer. */
```
fixed typos, thanks

kapi-no commented:
Last force push: fixed typo in the test according to @alwa-nordic suggestion

kapi-no commented:
Last force push: rebase on top of main; no diff has changed.

carlescufi commented:
This PR is currently being discussed internally at Nordic and may be dropped.

jori-nordic commented on subsys/bluetooth/host/smp.c:
```diff
@@ -3875,8 +3889,15 @@ static uint8_t smp_ident_addr_info(struct bt_smp *smp, struct net_buf *buf)
+					.id_addr = &req->addr,
+				};
+
+				bt_conn_foreach(BT_CONN_TYPE_ALL,
```
can you change that to `BT_CONN_TYPE_LE`? Otherwise you access the wrong union variant.

kapi-no replied:
thanks, it is fixed now

kapi-no commented:
Rebased my PR (no diff change)
